src/configuration.h

```diff
@@ -30,6 +30,9 @@ class MetadataAgentConfiguration {
   // value means that parsing succeeded, but all of the arguments were handled
   // within the function and the program should exit with a success exit code.
   int ParseArguments(int ac, char** av);
+  void ParseConfigurationStream(std::istream& input);
```

Comment:
Instead, can we please just add a public constructor with a `std::istream&` parameter? Something like:
```cpp
  MetadataAgentConfiguration::MetadataAgentConfiguration(std::istream& input)
      : MetadataAgentConfiguration() {
    ParseConfiguration(input);
  }
```
and then construct the stream in the tests as:
```cpp
  MetadataAgentConfiguration config(std::istringstream(
      "ProjectId: TestProjectId\n"
      "MetadataApiNumThreads: 13\n"
      "MetadataReporterPurgeDeleted: true\n"
      "MetadataReporterUserAgent: \"foobar/foobaz\"\n"
  ));
```
?

Reply:
@igorpeshansky would using a constructor instead of a public method still allow us to dynamically reload configurations in the future? It would be nice if the metadata agent could pick up changes to the configuration on the disk and reload in the future, either via watching for changes on the file itself periodically, or by receiving a signal to reload the configuration. This could help us reduce the current negative impact of the metadata agent becoming unavailable during a restart, if the only intended change is a configuration update. Kubernetes already updates the files on disk when they're updated via ConfigMaps or Secrets, which would make this an awesome candidate. If you feel that a constructor / public method is just a matter of implementation, fine by me, as long as we are open to the possibility of dynamically reloading the configuration.

Reply:
Adding to this, I know both "allow us", but what are your thoughts on the topic as a whole, and which approach would be more idiomatic in your mind?

Reply:
I believe that such a functionality, when it does get added, should be encapsulated in the `Configuration` class. No external entity should be responsible for keeping the configuration up to date.

Reply:
@igorpeshansky I agree with that principle, however its not represented in the current implementation. Right now the configuration class has an empty constructor and requires the something external to call `ParseArguments(int ac, char** av)`.

Perhaps we should have 2 constructors: `MetadataAgentConfiguration(std::istream& input)` and `MetadataAgentConfiguration(int ac, char ** av)`?

Reply:
I see, the problem is with error reporting.  `ParseArguments(int ac, char** av)` needs to return the success mode to the caller.  It does seem that this initialization needs to be done outside a constructor.

In the interest of moving things forward, I am going to implement the constructor and test helper as @igorpeshansky has described. Lets talk offline so I can better understand why this is the way to do it.

Reply:
Offline conclusion: We will make `int ParseArguments(int ac, char** av)` private, and `int ::main(int ac, char** av);` a friend of MetadataAgentConfiguration.  This friendship indicates an encapsulation leak that will be refactored out at a later point.